x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceAggregator.java

```diff
@@ -0,0 +1,169 @@
+        // context. The {@link DelegatingCircuitBreakerService} workarounds several accounting problems. Please have a look
+        // into the description of {@link DelegatingCircuitBreakerService} for more details.
+        this.breakerService = new DelegatingCircuitBreakerService(context.breaker(), this::addRequestCircuitBreakerBytes);
+        this.bigArraysForMapReduce = BigArrays.NON_RECYCLING_INSTANCE.withBreakerService(breakerService).withCircuitBreaking();
+        this.mapReduceContextByBucketOrdinal = new LongObjectPagedHashMap<>(1, context.bigArrays());
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalMapReduceAggregation<>(name, metadata(), mapReducer, null, null, fieldNames, profiling);
+    }
+
+    @Override
+    public final InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        InternalAggregation[] results = new InternalAggregation[owningBucketOrds.length];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            results[ordIdx] = buildAggregation(ordIdx);
+        }
+
+        return results;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return new LeafBucketCollectorBase(sub, null) {
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                SetOnce<IOException> firstException = new SetOnce<>();
+
+                mapReducer.map(extractors.stream().map(extractor -> {
+                    try {
+                        return extractor.collect(ctx, doc);
+                    } catch (IOException e) {
+                        firstException.trySet(e);
+                        // ignored in AbstractMapReducer
+                        return null;
+                    }
+                }), getMapReduceContext(owningBucketOrd));
+
+                if (firstException.get() != null) {
+                    throw firstException.get();
+                }
+            }
+        };
+    }
+
+    @Override
+    public void doPostCollection() {
+        for (long ordIdx = 0; ordIdx < mapReduceContextByBucketOrdinal.size(); ordIdx++) {
+            MapContext context = getMapReduceContext(ordIdx);
+            mapReduceContextByBucketOrdinal.put(ordIdx, mapReducer.mapFinalize(context));
+        }
+    }
+
+    @Override
+    public void collectDebugInfo(BiConsumer<String, Object> add) {
+        add.accept("map_reducer", mapReducer.getWriteableName());
+        mapReducer.collectDebugInfo(add);
+    }
+
+    @Override
+    protected void doClose() {
+        // disconnect the aggregation context circuit breaker, so big arrays used in results can be passed
+        if (breakerService != null) {
+            breakerService.disconnect();
```

Comment:
Okay.  I understand why we're not returning the bytes from the map reduce context to the CB here, but I don't see where they are being returned.  `InternalMapReduceAggregation` talks about how we manage the memory around deserialization, but I don't see where the ones created data node side release the memory they're holding from here.

Reply:
On the Aggregator side the `AggContext` takes care of circuit breaking, the bytes get reported to the `AggContext` and the `AggContext` returns the allocation to the breaker in `AggContext::close()`. That means the bytes are returned _before_ the actual object is destroyed. Between the AggContext getting closed and the internal agg getting send off, the object is allocated without proper accounting by a circuit breaker. That's the first problem described in #88128.

On the coordinating side the memory gets accounted in `QueryPhaseResultConsumer::estimateRamBytesUsedForReduce`, that's the infamous `1.5` magic. That means on the reduce side we simply don't do any CB magic and assume it fits in the budget. `QueryPhaseResultConsumer` allocates the budget and releases it for us. Most probably it over-allocates,  problems 2-5 listed in #88128 talk about that part.

LBNL the big arrays aren't using recycling. I don't think I can guarantee returning the buffers properly in all error scenarios I can think of.